File: go-version/internal/retriever/retriever.go
// Package retriever implements the SAAI Expert Knowledge Retriever.
//
// It loads all expert methodology documents from the /docs/experts directory,
// chunks them into passages, and exposes a Query interface so agents can
// retrieve the most relevant expert knowledge for a given SEO task.
//
// Architecture:
//
//	ExpertDoc → []Chunk → InMemoryStore → BM25Retriever
//	                                    ↕
//	                              SemanticRetriever (optional, via embeddings)
package retriever

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// ─── Types ────────────────────────────────────────────────────────────────────

// ExpertDoc is a fully loaded expert methodology document.
type ExpertDoc struct {
	ExpertName string // e.g. "Koray Tuğberk GÜBÜR"
	AgentID    string // e.g. "SemanticTopicalAuthorityAgent"
	FilePath   string
	Content    string
}

// Chunk is a paragraph-level passage from an ExpertDoc.
type Chunk struct {
	ExpertName string
	AgentID    string
	Text       string
	Source     string // filepath + heading context
}

// Result is a retrieved chunk with a relevance score.
type Result struct {
	Chunk
	Score float64
}

// ─── ExpertStore ─────────────────────────────────────────────────────────────

// ExpertStore holds all loaded expert documents and their derived chunks.
type ExpertStore struct {
	Docs   []ExpertDoc
	Chunks []Chunk
}

// expertAgentMap maps expert markdown file stems to their AgentIDs.
var expertAgentMap = map[string]string{
	"koray-tugberk-gubur": "SemanticTopicalAuthorityAgent",
	"jamie-indigo":        "JSRenderingCrawlAgent",
	"jono-alderson":       "TechnicalSpeedAgent",
	"lily-ray":            "EEATContentAgent",
	"patrick-stox":        "OnPageLogAnalysisAgent",
	"marie-haynes":        "GoogleUpdateAgent",
	"aleyda-solis":        "InternationalSEOAgent",
	"barry-adams":         "NewsSEOAgent",
}

// LoadExpertStore reads all markdown files from docsDir and builds the store.
// docsDir should point to the /docs/experts directory.
func LoadExpertStore(docsDir string) (*ExpertStore, error) {
	entries, err := os.ReadDir(docsDir)
	if err != nil {
		return nil, fmt.Errorf("retriever: cannot read docs dir %q: %w", docsDir, err)
	}

	store := &ExpertStore{}

	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}

		filePath := filepath.Join(docsDir, entry.Name())
		content, readErr := os.ReadFile(filePath)
		if readErr != nil {
			return nil, fmt.Errorf("retriever: reading %q: %w", filePath, readErr)
		}

		stem := strings.TrimSuffix(entry.Name(), ".md")
		agentID, ok := expertAgentMap[stem]
		if !ok {
			agentID = "UnknownAgent"
		}

		// Extract expert name from first H1 heading
		expertName := extractH1(string(content))
		if expertName == "" {
			expertName = stem
		}

		doc := ExpertDoc{
			ExpertName: expertName,
			AgentID:    agentID,
			FilePath:   filePath,
			Content:    string(content),
		}
		store.Docs = append(store.Docs, doc)

		// Chunk the document into paragraphs
		chunks := chunkDocument(doc)
		store.Chunks = append(store.Chunks, chunks...)
	}

	return store, nil
}

// ─── Chunking ─────────────────────────────────────────────────────────────────

// chunkDocument splits an ExpertDoc into non-empty paragraph chunks.
func chunkDocument(doc ExpertDoc) []Chunk {
	var chunks []Chunk
	currentHeading := ""

	scanner := bufio.NewScanner(strings.NewReader(doc.Content))
	var paragraphLines []string

	flush := func() {
		text := strings.TrimSpace(strings.Join(paragraphLines, "\n"))
		if len(text) > 30 { // ignore tiny fragments
			chunks = append(chunks, Chunk{
				ExpertName: doc.ExpertName,
				AgentID:    doc.AgentID,
				Text:       text,
				Source:     doc.FilePath + " → " + currentHeading,
			})
		}
		paragraphLines = nil
	}

	for scanner.Scan() {
		line := scanner.Text()

		// Track current heading context
		if strings.HasPrefix(line, "#") {
			flush()
			currentHeading = strings.TrimLeft(line, "# ")
			continue
		}

		// Blank line = paragraph boundary
		if strings.TrimSpace(line) == "" {
			flush()
			continue
		}

		paragraphLines = append(paragraphLines, line)
	}
	flush()

	return chunks
}

// extractH1 returns the text of the first H1 heading in a markdown document.
func extractH1(content string) string {
	for _, line := range strings.Split(content, "\n") {
		if strings.HasPrefix(line, "# ") {
			return strings.TrimPrefix(line, "# ")
		}
	}
	return ""
}

// ─── BM25 Retriever ───────────────────────────────────────────────────────────

// Retriever provides BM25-based lexical search over the expert knowledge base.
// BM25 constants (Okapi BM25)
const (
	bm25K1 = 1.5
	bm25B  = 0.75
)

// Retriever wraps an ExpertStore with BM25 scoring capability.
type Retriever struct {
	store       *ExpertStore
	tf          []map[string]float64 // term frequency per chunk
	df          map[string]int       // document frequency per term
	avgLen      float64              // average chunk length (in words)
}

// New creates a Retriever from a loaded ExpertStore and pre-computes BM25 stats.
func New(store *ExpertStore) *Retriever {
	r := &Retriever{
		store: store,
		df:    make(map[string]int),
	}

	var totalLen float64
	r.tf = make([]map[string]float64, len(store.Chunks))

	for i, chunk := range store.Chunks {
		tokens := tokenize(chunk.Text)
		totalLen += float64(len(tokens))
		freq := make(map[string]float64)
		seen := make(map[string]bool)

		for _, tok := range tokens {
			freq[tok]++
			if !seen[tok] {
				r.df[tok]++
				seen[tok] = true
			}
		}
		r.tf[i] = freq
	}

	if len(store.Chunks) > 0 {
		r.avgLen = totalLen / float64(len(store.Chunks))
	}

	return r
}

// Query retrieves the top-k most relevant chunks for a natural language query.
func (r *Retriever) Query(query string, topK int) []Result {
	queryTokens := tokenize(query)
	N := float64(len(r.store.Chunks))

	scores := make([]float64, len(r.store.Chunks))

	for _, qt := range queryTokens {
		df := float64(r.df[qt])
		if df == 0 {
			continue
		}
		// IDF with smoothing
		idf := math.Log((N-df+0.5)/(df+0.5) + 1)

		for i, freq := range r.tf {
			tf := freq[qt]
			if tf == 0 {
				continue
			}
			docLen := float64(len(tokenize(r.store.Chunks[i].Text)))
			norm := 1 - bm25B + bm25B*(docLen/r.avgLen)
			scores[i] += idf * ((tf * (bm25K1 + 1)) / (tf + bm25K1*norm))
		}
	}

	// Build and sort results
	results := make([]Result, len(r.store.Chunks))
	for i, chunk := range r.store.Chunks {
		results[i] = Result{Chunk: chunk, Score: scores[i]}
	}
	sort.Slice(results, func(i, j int) bool {
		return results[i].Score > results[j].Score
	})

	if topK > len(results) {
		topK = len(results)
	}
	// Filter zero-score results
	filtered := results[:0]
	for _, r := range results[:topK] {
		if r.Score > 0 {
			filtered = append(filtered, r)
		}
	}
	return filtered
}

// QueryByExpert retrieves the top-k chunks from a specific expert's documents.
func (r *Retriever) QueryByExpert(query, agentID string, topK int) []Result {
	all := r.Query(query, len(r.store.Chunks))
	var filtered []Result
	for _, res := range all {
		if res.AgentID == agentID {
			filtered = append(filtered, res)
			if len(filtered) >= topK {
				break
			}
		}
	}
	return filtered
}

// ListExperts returns a summary of all loaded expert documents.
func (r *Retriever) ListExperts() []ExpertDoc {
	return r.store.Docs
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

// tokenize lowercases and splits text into terms, stripping markdown punctuation.
func tokenize(text string) []string {
	var tokens []string
	text = strings.ToLower(text)

	// Replace common markdown/punctuation with spaces
	replacer := strings.NewReplacer(
		"#", " ", "*", " ", "`", " ", "_", " ",
		".", " ", ",", " ", ":", " ", ";", " ",
		"(", " ", ")", " ", "[", " ", "]", " ",
		"\n", " ", "\t", " ", "\"", " ", "'", " ",
		"-", " ", "/", " ", "\\", " ", "|", " ",
	)
	text = replacer.Replace(text)

	for _, tok := range strings.Fields(text) {
		if len(tok) > 2 { // skip very short tokens
			tokens = append(tokens, tok)
		}
	}
	return tokens
}

File: kotlin-version/src/main/kotlin/retriever/Retriever.kt
package retriever

import java.io.File
import kotlin.math.ln
import kotlin.math.sqrt

// ─── Data Models ─────────────────────────────────────────────────────────────

/** A fully loaded expert methodology document. */
data class ExpertDoc(
    val expertName: String,
    val agentId: String,
    val filePath: String,
    val content: String
)

/** A paragraph-level passage derived from an [ExpertDoc]. */
data class Chunk(
    val expertName: String,
    val agentId: String,
    val text: String,
    val source: String       // filePath → heading context
)

/** A retrieved [Chunk] alongside its BM25 relevance score. */
data class Result(
    val chunk: Chunk,
    val score: Double
)

// ─── Expert → Agent Mapping ───────────────────────────────────────────────────

private val EXPERT_AGENT_MAP = mapOf(
    "koray-tugberk-gubur" to "SemanticTopicalAuthorityAgent",
    "jamie-indigo"        to "JSRenderingCrawlAgent",
    "jono-alderson"       to "TechnicalSpeedAgent",
    "lily-ray"            to "EEATContentAgent",
    "patrick-stox"        to "OnPageLogAnalysisAgent",
    "marie-haynes"        to "GoogleUpdateAgent",
    "aleyda-solis"        to "InternationalSEOAgent",
    "barry-adams"         to "NewsSEOAgent",
)

// ─── ExpertStore ─────────────────────────────────────────────────────────────

/**
 * Loads all `*.md` files from [docsDir] and chunks them into paragraphs,
 * building the in-memory knowledge base used by the BM25 [Retriever].
 */
class ExpertStore(private val docsDir: File) {

    val docs: List<ExpertDoc>
    val chunks: List<Chunk>

    init {
        require(docsDir.isDirectory) {
            "ExpertStore: '$docsDir' is not a directory"
        }

        val loadedDocs = docsDir
            .listFiles { f -> f.extension == "md" }
            ?.sortedBy { it.name }
            ?.map { file ->
                val stem    = file.nameWithoutExtension
                val content = file.readText()
                ExpertDoc(
                    expertName = extractH1(content) ?: stem,
                    agentId    = EXPERT_AGENT_MAP[stem] ?: "UnknownAgent",
                    filePath   = file.absolutePath,
                    content    = content
                )
            }
            ?: emptyList()

        docs   = loadedDocs
        chunks = loadedDocs.flatMap { chunkDocument(it) }
    }

    // ── Chunking ──────────────────────────────────────────────────────────────

    private fun chunkDocument(doc: ExpertDoc): List<Chunk> {
        val result      = mutableListOf<Chunk>()
        var heading     = ""
        val lines       = mutableListOf<String>()

        fun flush() {
            val text = lines.joinToString("\n").trim()
            if (text.length > 30) {
                result += Chunk(
                    expertName = doc.expertName,
                    agentId    = doc.agentId,
                    text       = text,
                    source     = "${doc.filePath} → $heading"
                )
            }
            lines.clear()
        }

        for (line in doc.content.lines()) {
            when {
                line.startsWith("#") -> {
                    flush()
                    heading = line.trimStart('#', ' ')
                }
                line.isBlank() -> flush()
                else           -> lines += line
            }
        }
        flush()

        return result
    }

    private fun extractH1(content: String): String? =
        content.lines()
            .firstOrNull { it.startsWith("# ") }
            ?.removePrefix("# ")
            ?.trim()
}

// ─── BM25 Retriever ───────────────────────────────────────────────────────────

private const val BM25_K1 = 1.5
private const val BM25_B  = 0.75

/**
 * Okapi BM25 retriever over an [ExpertStore].
 *
 * On construction it pre-computes term frequencies (TF),
 * document frequencies (DF), and average document length
 * so that [query] runs in O(|queryTerms| × |chunks|) time.
 */
class Retriever(private val store: ExpertStore) {

    private val tf:     List<Map<String, Double>>   // TF per chunk
    private val df:     Map<String, Int>            // DF per term
    private val avgLen: Double                      // avg chunk length in tokens

    init {
        val tfMutable = mutableListOf<Map<String, Double>>()
        val dfMutable = mutableMapOf<String, Int>()
        var totalLen  = 0.0

        for (chunk in store.chunks) {
            val tokens = tokenize(chunk.text)
            totalLen  += tokens.size
            val freq   = mutableMapOf<String, Double>()
            val seen   = mutableSetOf<String>()

            for (tok in tokens) {
                freq[tok] = (freq[tok] ?: 0.0) + 1.0
                if (seen.add(tok)) dfMutable[tok] = (dfMutable[tok] ?: 0) + 1
            }
            tfMutable += freq
        }

        tf     = tfMutable
        df     = dfMutable
        avgLen = if (store.chunks.isNotEmpty()) totalLen / store.chunks.size else 1.0
    }

    /**
     * Returns the top-[k] most relevant [Result]s for a natural-language [query].
     */
    fun query(query: String, k: Int = 5): List<Result> {
        val queryTokens = tokenize(query)
        val n           = store.chunks.size.toDouble()
        val scores      = DoubleArray(store.chunks.size)

        for (qt in queryTokens) {
            val docFreq = df[qt]?.toDouble() ?: continue
            val idf     = ln((n - docFreq + 0.5) / (docFreq + 0.5) + 1.0)

            for ((i, chunkTf) in tf.withIndex()) {
                val termFreq = chunkTf[qt] ?: continue
                val docLen   = tokenize(store.chunks[i].text).size.toDouble()
                val norm     = 1 - BM25_B + BM25_B * (docLen / avgLen)
                scores[i]   += idf * ((termFreq * (BM25_K1 + 1)) / (termFreq + BM25_K1 * norm))
            }
        }

        return store.chunks.indices
            .map { Result(store.chunks[it], scores[it]) }
            .filter { it.score > 0 }
            .sortedByDescending { it.score }
            .take(k)
    }

    /**
     * Retrieves top-[k] results restricted to a single expert's documents.
     */
    fun queryByExpert(query: String, agentId: String, k: Int = 3): List<Result> =
        query(query, store.chunks.size)
            .filter { it.chunk.agentId == agentId }
            .take(k)

    /** Returns all loaded [ExpertDoc]s for display / debug. */
    fun listExperts(): List<ExpertDoc> = store.docs
}

// ─── Tokeniser ───────────────────────────────────────────────────────────────

private val STRIP_REGEX = Regex("""[#*`_.,;:()\[\]"\n\t'\\|/\-]+""")

private fun tokenize(text: String): List<String> =
    STRIP_REGEX.replace(text.lowercase(), " ")
        .split(" ")
        .filter { it.length > 2 }
